cloud_functions/gcs_event_based_ingest/gcs_ocn_bq_ingest/main.py

```diff
@@ -0,0 +1,378 @@
+    gcs = storage.Client(client_info=client_info)
+    default_query_config = bigquery.QueryJobConfig()
+    default_query_config.use_legacy_sql = False
+    default_query_config.labels = DEFAULT_JOB_LABELS
+    bq = bigquery.Client(
+        client_info=client_info, default_query_job_config=default_query_config
+    )
+
+    gsurl = f"gs://{bucket_id}/{prefix_to_load}"
+
+    logging.debug(f"looking for {gsurl}_config/bq_transform.sql")
+    print(f"looking for {gsurl}_config/bq_transform.sql")
+    external_query_sql = read_gcs_file_if_exists(gcs, f"{gsurl}_config/bq_transform.sql")
+    logging.debug(f"external_query_sql = '{external_query_sql}'")
+    print(f"external_query_sql = {external_query_sql}")
+    if external_query_sql:
+        logging.debug("EXTERNAL QUERY")
+        print("EXTERNAL QUERY")
+        external_query(gcs, bq, gsurl, external_query_sql, dest_table_ref)
+        return
+
+    logging.debug("LOAD_JOB")
+    print("LOAD_JOB")
+    load_batches(gcs, bq, gsurl, dest_table_ref)
+
+
+def external_query(gcs, bq, gsurl, query, dest_table_ref):
+    """Load from query over external table from GCS.
+
+    This hinges on a SQL query defined in GCS at _config/bq_transform.sql and
+    an external table definition _config/external.json (otherwise will assume
+    parquet external table)
+    """
+    external_table_config = read_gcs_file_if_exists(gcs, f"{gsurl}_config/external.json")
+    logging.debug("reading external table config")
+    if external_table_config:
+        external_table_def = json.loads(external_table_config)
+    else:
+        print(f"Falling back to default parquet external table."
+              f" {gsurl}/_config/external.json not found.")
+        external_table_def = DEFAULT_EXTERNAL_TABLE_DEFINITION
+
+    external_table_def["sourceUris"] = flatten(get_batches_for_prefix(gcs, gsurl))
+    external_config = bigquery.ExternalConfig.from_api_repr(
+        external_table_def)
+    job_config = bigquery.QueryJobConfig(
+        table_definitions={"temp_ext": external_config},
+        use_legacy_sql=False
+    )
+    # for some reason string literal wrapped in b''
+    rendered_query = str(str(query).format(
+        dest_dataset=dest_table_ref.dataset_id,
+        dest_table=dest_table_ref.table_id))[2:-1]
+
+    job: bigquery.QueryJob = bq.query(
+        rendered_query,
+        job_config=job_config)
+
+    print(f"started asynchronous query job: {job.job_id}")
+
+    start_poll_for_errors = monotonic()
+    # Check if job failed quickly
+    while monotonic() - start_poll_for_errors < WAIT_FOR_JOB_SECONDS:
+        job.reload()
+        if job.errors:
+            raise RuntimeError(
+                f"query job {job.job_id} failed quickly: {job.errors}")
+        sleep(1)
+
+
+def flatten(arr: List[List[Any]]) -> List[Any]:
+    """Flatten list of lists to flat list of elements"""
+    return [j for i in arr for j in i]
+
+
+def load_batches(gcs, bq, gsurl, dest_table_ref):
```

Comment:
without changing the configuration interface we could optionally swap out this `load_batches` for creating a GCS to BQ Data Transfer Service config and submitting a new data transfer. This would have the advantage of relying on DTS's existing mechanics to orchestrate multiple loads to a staging table and atomically commit partitions to the destination table with a copy job. This has the advantage of atomicity of loading a partition but the disadvantage of a requiring a call to bigquerydatatransfer.googleapis.com which may prohibit use in environments that can't route this address due to lack of restricted VIP.

Reply:
I think you should still be able to use BigQuery DTS GCS to BQ inside the VPC-SC. Not sure where the gaps are between transfer services but other customers have been running these types of jobs inside the VPC.